Enable userspace mode on ARM64

npitre opened the pull request:
Remaining pieces to tie userspace support together.
Mostly trivias.

carlocaione committed:
tests: userspace: Add AArch64 case

Fix tests/kernel/mem_protect/userspace test adding the arch-specific
code to support AArch64.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
tests: userspace: Fix test for AArch64

Fix tests/kernel/mem_protect/mem_protect/ adding support for AArch64.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
aarch64: userspace: Increase KOBJECT_TEXT_AREA

This is needed to have some tests run successfully.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
tests: userspace: Increase mem pool size for AArch64

Fix tests/kernel/threads/dynamic_thread/ increasing the mem pool size.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
aarch64: userspace: Add support for page tables swapping

Introduce the necessary routines to have the user thread stack correctly
mapped and the functions to swap page tables on context switch.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
aarch64: mmu: Fix MMU permissions for zephyr code and data

User threads still need to access the code and the RO data. Fix the
permissions.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
aarch64: userspace: Enable userspace

Add ARCH_HAS_USERSPACE to enable userspace.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
aarch64: userspace: Add missing entries in vector table

To support exceptions taken in EL0.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
aarch64: GCC10: Add -mno-outline-atomics

GCC10 introduced by default calls to out-of-line helpers to implement
atomic operations with the '-moutline-atomic' option. This is breaking
several tests because the embedded calls are trying to access the
zephyr_data region from userspace that is declared as MT_P_RW_U_NA,
triggering a memory fault.

Since there is currently no support for MT_P_RW_U_RO (and probably never
will be), disable the out-of-line helpers disabling the GCC option.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

npitre committed:
aarch64: userspace: bump the global number of available page tables

Each memory domain requires a few pages for itself.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

dcpleung commented on arch/arm/core/aarch64/Kconfig:
```diff
@@ -67,6 +67,9 @@ config AARCH64_IMAGE_HEADER
 	default 4096
 
+config KOBJECT_TEXT_AREA
+	default 512
```
If this is simply for test and not needed in general, might be a good idea to make it conditional (i.e. `if TEST`)

carlocaione replied:
ack and fixed.

npitre committed:
aarch64: mmu: apply domain switching on all CPUs if SMP

It is apparently possible for one CPU to change the memory domain
of a thread already being executed on another CPU.

All CPUs must ensure they're using the appropriate mapping after a
thread is newly added to a domain.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>
